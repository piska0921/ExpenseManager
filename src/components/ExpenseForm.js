import React from 'react'
import PropTypes from 'prop-types'
import { validateDate } from '../utility'

class ExpenseForm extends React.Component {
    constructor(props) {
        super(props)
        this.state = {
            editItem: props.editItem,
            isValidated: true,
            errorMsg: '',
            isEdit: !(props.editItem && Object.keys(props.editItem).length === 0)
        }
    }

    submitForm = (event) => {
        const isEdit = this.state.isEdit
        const title = this.titleInput.value.trim()
        const amount = this.amountInput.value.trim() * 1
        const date = this.dateInput.value.trim()
        event.preventDefault()
        if (title && amount && date) {
            if (amount <= 0) {
                this.setState({
                    isValidated: false,
                    errorMsg: 'Amount must be greater than 0'
                })
            } else if (!validateDate(date)) {
                this.setState({
                    isValidated: false,
                    errorMsg: 'Please enter correct date'
                })
            } else {
                if (!isEdit) {
                    const data = { title, amount, date }
                    this.props.onFormSubmit(data, isEdit)
                } else {
                    delete this.props.editItem.category
                    this.props.onFormSubmit({...this.props.editItem, title, amount, date}, isEdit)
                }
            }
        } else {
            this.setState({
                isValidated: false,
                errorMsg: 'Please fill in all required fields'
            })
        }

    }
    render() {
        const {editItem} = this.props
        return (
            <form style={{ textAlign: 'left' }} onSubmit={(event) => { this.submitForm(event) }} noValidate>
                <div className="form-group">
                    <label htmlFor="title">Title</label>
                    <input
                        type="text" className="form-control"
                        id="title" placeholder="Please enter title"
                        defaultValue={(Object.keys(editItem).length === 0) ? '' : editItem.title}
                        ref={(input) => { this.titleInput = input }}
                    />
                </div>
                <div className="form-group">
                    <label htmlFor="amount">Amount</label>
                    <div className="input-group">
                        <div className="input-group-prepend">
                            <span className="input-group-text">$</span>
                        </div>
                        <input
                            type="number" className="form-control"
                            id="amount" placeholder="Please enter amount"
                            defaultValue={(Object.keys(editItem).length === 0) ? '' : editItem.amount}
                            ref={(input) => { this.amountInput = input }}
                        />
                    </div>
                </div>
                <div className="form-group">
                    <label htmlFor="date">Date</label>
                    <input
                        type="date" className="form-control"
                        id="date" placeholder="Please choose date"
                        defaultValue={(Object.keys(editItem).length === 0) ? '' : editItem.date}
                        ref={(input) => { this.dateInput = input }}
                    />
                </div>
                <button type="submit" className="btn btn-primary mr-3">Save</button>
                <button className="btn btn-secondary" onClick={this.props.onFormCancel}>Cancel</button>
                { !this.state.isValidated && <div className="alert alert-danger mt-5" role="alert">{this.state.errorMsg}</div>}
            </form>

        )
    }
}

ExpenseForm.propTypes = {
    onFormSubmit: PropTypes.func.isRequired,
    onFormCancel: PropTypes.func.isRequired,
    editItem: PropTypes.object
}

ExpenseForm.defaultProps = {
    editItem: {}
}

export default ExpenseForm