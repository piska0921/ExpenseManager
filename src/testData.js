export const testCategories = [
    {
        "id": "1",
        "name": "travel",
        "type": "expense",
        "iconName": "ios-plane"
    },
    {
        "id": "2",
        "name": "finance",
        "type": "expense",
        "iconName": "logo-yen"
    },
    {
        "id": "3",
        "name": "food",
        "type": "expense",
        "iconName": "ios-pizza"
    },
    {
        "id": "4",
        "name": "cloth",
        "type": "expense",
        "iconName": "ios-shirt"
    },
    {
        "id": "5",
        "name": "salary",
        "type": "income",
        "iconName": "ios-card"
    },
    {
        "id": "5",
        "name": "salary",
        "type": "income",
        "iconName": "ios-card"
    }

]

export const testItems = [
    {
        "id": "1",
        "title": "travel to New Zealand",
        "date": "2020-09-11",
        "amount": 2300,
        "categoryId": "1"
    },
    {
        "id": "2",
        "title": "Purchase Stock",
        "date": "2020-09-11",
        "amount": 500,
        "categoryId": "2"
    }
]