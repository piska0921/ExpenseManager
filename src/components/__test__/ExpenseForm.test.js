import React from 'react'
import { mount } from 'enzyme'
import ExpenseForm from '../ExpenseForm'
import { testItems } from '../../testData'

const props = {
    onFormSubmit: jest.fn(),
    onFormCancel: jest.fn(),
    editItem: {}
}

const props_with_editItem = {
    onFormSubmit: jest.fn(),
    onFormCancel: jest.fn(),
    editItem: testItems[0]
}

const getInputValue = (wrapper, selector) => {
    return wrapper.find(selector).instance().value
}
const setInputValue = (wrapper, selector, input) => {
    wrapper.find(selector).instance().value = input
}
let wrapper, wrapper2, formInstance
describe('test expense form component', () => {
    beforeEach(() => {
        wrapper = mount(<ExpenseForm {...props} />)
        wrapper2 = mount(<ExpenseForm {...props_with_editItem} />)
        formInstance = wrapper.find(ExpenseForm).instance()
    })
    it('should render the component to match the snapshot', () => {
        expect(wrapper).toMatchSnapshot()
        expect(wrapper2).toMatchSnapshot()
    })
    describe('should correctly render the component with no editItem', () => {
        it('should render three input boxes, two buttons and one form', () => {
            expect(wrapper.find('input').length).toEqual(3)
            expect(wrapper.find('button').length).toEqual(2)
            expect(wrapper.find('form').length).toEqual(1)
        })
        it('should render input boxes with empty default value', () => {
            expect(getInputValue(wrapper, '#title')).toEqual('')
            expect(getInputValue(wrapper, '#amount')).toEqual('')
            expect(getInputValue(wrapper, '#date')).toEqual('')
        })
        it('submit form with empty input should give alert message and not trigger submit action', () => {
            wrapper.find('form').simulate('submit')
            expect(wrapper.find('.alert').length).toEqual(1)
            expect(formInstance.state.isValidated).toEqual(false)
            expect(props.onFormSubmit).not.toHaveBeenCalled()
        })
        it('submit form with negative amount should give alert message and not trigger submit action', () => {
            setInputValue(wrapper, '#title', 'title input')
            setInputValue(wrapper, '#amount', '-10')
            setInputValue(wrapper, '#date', '2020-12-20')
            wrapper.find('form').simulate('submit')
            expect(wrapper.find('.alert').length).toEqual(1)
            expect(formInstance.state.isValidated).toEqual(false)
            expect(props.onFormSubmit).not.toHaveBeenCalled()
        })
        it('submit form with invalid date should give alert message and not trigger submit action', () => {
            setInputValue(wrapper, '#title', 'title input')
            setInputValue(wrapper, '#amount', '10')
            setInputValue(wrapper, '#date', 'invalid date')
            wrapper.find('form').simulate('submit')
            expect(wrapper.find('.alert').length).toEqual(1)
            expect(formInstance.state.isValidated).toEqual(false)
            expect(props.onFormSubmit).not.toHaveBeenCalled()
        })
        it('submit form with correct inputs should trigger submit action', () => {
            setInputValue(wrapper, '#title', 'title input')
            setInputValue(wrapper, '#amount', '10')
            setInputValue(wrapper, '#date', '2020-12-20')
            wrapper.find('form').simulate('submit')
            const inputItem = {
                title: 'title input',
                amount: 10,
                date: '2020-12-20'
            }
            expect(props.onFormSubmit).toHaveBeenCalledWith(inputItem, false)
        })
        it('click cancel button should trigger cancel action', () => {
            wrapper.find('button').last().simulate('click')
            expect(props.onFormCancel).toHaveBeenCalled()
        })
    })
    describe('should render the component correctly with editItem', () => {
        it('input boxes should be prepopulated with editItem data', () => {
            expect(getInputValue(wrapper2, '#title')).toEqual(testItems[0].title)
            expect(getInputValue(wrapper2, '#amount')).toEqual(testItems[0].amount.toString())
            expect(getInputValue(wrapper2, '#date')).toEqual(testItems[0].date)
        })
        it('submit form with changed value should trigger submit action with correct item', () => {
            setInputValue(wrapper2, '#title', 'new title')
            wrapper2.find('form').simulate('submit')
            const newItem = { ...testItems[0], title: 'new title'}
            expect(props_with_editItem.onFormSubmit).toHaveBeenCalledWith(newItem, true)
        })
    })

})