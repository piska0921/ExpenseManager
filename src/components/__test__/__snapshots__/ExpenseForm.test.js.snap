// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test expense form component should render the component to match the snapshot 1`] = `ReactWrapper {}`;

exports[`test expense form component should render the component to match the snapshot 2`] = `ReactWrapper {}`;
